velox/functions/lib/string/StringImpl.h

```diff
@@ -162,6 +162,105 @@ FOLLY_ALWAYS_INLINE int32_t charToCodePoint(const T& inputString) {
   return codePoint;
 }
 
+template <bool isAscii, typename TInString>
+FOLLY_ALWAYS_INLINE int64_t
+hammingDistance(const TInString& left, const TInString& right) {
+  int64_t distance = 0;
+  int64_t leftPosition = 0;
+  int64_t rightPosition = 0;
+  std::string_view leftSV = std::string_view(left.data(), left.size());
+  std::string_view rightSV = std::string_view(right.data(), right.size());
+
+  while (leftPosition < leftSV.size() && rightPosition < rightSV.size()) {
+    int leftSize, rightSize;
+    auto codePointLeft = isAscii
+        ? leftSV.at(leftPosition)
+        : utf8proc_codepoint(&leftSV.at(leftPosition), leftSize);
+    auto codePointRight = isAscii
+        ? rightSV.at(rightPosition)
+        : utf8proc_codepoint(&rightSV.at(rightPosition), rightSize);
+
+    // if both code points are invalid, the behavior is undefined
+    // the following code skips one byte and continues
+    leftPosition += codePointLeft > 0 ? leftSize : -codePointLeft;
+    rightPosition += codePointRight > 0 ? rightSize : -codePointRight;
+
+    if (codePointLeft != codePointRight) {
+      distance++;
+    }
+  }
+
+  VELOX_USER_CHECK(
+      leftPosition == leftSV.size() && rightPosition == rightSV.size(),
+      "The input strings to hamming_distance function must have the same length");
+
+  return distance;
+}
+
+template <bool isAscii>
+FOLLY_ALWAYS_INLINE void castToCodePoints(
+    std::string_view str,
+    std::vector<int32_t>& codePoints) {
+  for (int index = 0; index < str.size();) {
+    int size = 1;
+    auto codePoint =
+        isAscii ? str.at(index) : utf8proc_codepoint(&str.at(index), size);
+    VELOX_USER_CHECK(
+        codePoint > 0, "Invalid UTF-8 encoding in characters: {}", str.data());
+    index += size;
+    codePoints.push_back(codePoint);
+  }
+}
+
+template <bool isAscii, typename TInString>
+FOLLY_ALWAYS_INLINE int64_t
+levenshteinDistance(const TInString& left, const TInString& right) {
+  std::vector<int32_t> leftCodePoints, rightCodePoints;
+  castToCodePoints<isAscii>(
```

Comment:
It is necessary to allocate 2 vectors of code points? Would it be possible to work with one pair of code points at a time?

Reply:
The algorithm needs to compare every code point from one string to every code point in the other. This is done in the nested loop below. It will be possible to make this work using one vector (for the larger string), and extracting each code point from the smaller one for each iteration, but imo that would make the code look clunky.

Levenshtein distance is apparently[ proven to be of at least quadratic complexity](https://en.wikipedia.org/wiki/Levenshtein_distance#:~:text=Computational%20complexity%5Bedit%5D), although some approximate algorithms exist that have better time/space complexity.

Reply:
Thank you for checking and explaining. As written, 2 vectors and an array are allocated and freed for each row being processed. One way to reduce memory allocations would be to re-use vectors and array across different rows by turning them into member variables.

Reply:
It seems that we don't need the vector for the case when both strings are ASCII. In this case we can avoid memory allocations altogether. Since all-ASCII strings are very common, it would be nice to optimize for that case.